### PR TITLE
Remove device memory allocation out of memory pool

### PR DESCRIPTION
**Pull request by takagi**
This PR removes device memory allocation out of memory pool form `cupy.sort`.

### CODE REVIEW

File: cupy/cuda/cupy_thrust.cu [hunk 1]
```diff
@@ -13,56 +13,57 @@ using namespace thrust;
  */
 
 template <typename T>
-void cupy::thrust::_sort(void *start, const std::vector<ptrdiff_t>& shape) {
+void cupy::thrust::_sort(void *data_start, void *keys_start, const std::vector<ptrdiff_t>& shape) {
```
Comment (okuta) [inline cupy/cuda/cupy_thrust.cu]: How about `size_t *keys_start`?

File: cupy/cuda/cupy_thrust.cu [hunk 2]
```diff
@@ -13,56 +13,57 @@ using namespace thrust;
...
                   make_constant_iterator<ptrdiff_t>(shape[ndim-1]),
-                  d_keys.begin(),
+                  dp_keys_first,
                   divides<size_t>());
 
         // Sorting with back-to-back approach.
-        stable_sort_by_key(dp_first,
-                           dp_last,
-                           d_keys.begin(),
+        stable_sort_by_key(dp_data_first,
```
Comment (okuta) [inline cupy/cuda/cupy_thrust.cu]: How about using zip_iterator?

Comment (takagi) [inline cupy/cuda/cupy_thrust.cu]: Fixed!

File: cupy/cuda/cupy_thrust.cu (hunk 1 above)
Comment (takagi) [inline cupy/cuda/cupy_thrust.cu]: Fixed!

**Comment by okuta** [comment]
LGTM!

**Pull request merged**